accepted/future-releases/spread-collections/performance.md

```diff
@@ -0,0 +1,400 @@
+A key part of the spread proposal is the `...` syntax "desugars" to calls to a
+public API on the object being spread. This is important because it allows users
+to spread their own collection types. In other words, it makes the spread syntax
+programmable.
+
+That raises the question of *what* public API it should call. The API should be
+simple to implement, but also efficient to execute. Whatever it desugars to,
+we'll be stuck with forever. Since it's calling into user code which can have
+arbitrary side effects, any change to the semantics could be breaking.
+
+I came up with a handful of different desugarings and then wrote some, I hope,
+robust benchmarks. I tested them across a few platforms, collection sizes, and
+collection types. This produces something like a 5-dimensional hypercube of
+configurations, so I sliced the data a bunch of different ways and tried to
+summzarize the trends.
+
+**TL;DR: I recommend, we:**
+
+*   Desugar to `List.length` and `List.[]` at compile time if we statically know
+    the Iterable object being spread implements List.
+
+*   Desugar maps to `Map.keys` and `Map.[]`.
```

Comment:
Unless we can encourage all map implementations to make this *efficient* (not do an extra lookup), I really don't want to make it the mandatory approach.